core/src/main/scala/cats/free/Codensity.scala

```diff
@@ -0,0 +1,64 @@
+package cats
+package free
+
+/** 
+ * Codensity Monad
+ * 
+ * Codensity is used to improve quadratic complexity of free monadic structures.
+ * This approach is presented in this paper:
+ * http://www.janis-voigtlaender.eu/papers/AsymptoticImprovementOfComputationsOverFreeMonads.pdf
+ *
+ * The idea behind Codensity is simply to replace left associated structures by wrapping into right associated codensity.
+ * There is an isomorphism between Free monadic structure and codensity monad making it lossless to be used in your code.
+ *
+ * Interesting aspects to know:  
+ * - Codensity[F[_], A] = NaturalTransformation[λ[α => Kleisli[F, A, α]], F]
+ * - For any type constructor F[_], Codensity is a monad
+ */
+trait Codensity[F[_], A] {
```

Comment:
I have seen a tendency to replace trait by abstract class when we are not usint it as a mixin. It might be for performance reason but tbh I am not sure